File: apps/desktop/src/utils/providers.ts
import type { ApiProvider } from "../api";

export type ProviderCapability = "llm" | "asr";

export function providerCapability(provider: ApiProvider): ProviderCapability {
  if (provider.capability === "asr" || provider.capability === "llm") return provider.capability;
  if ((provider.transcription_model || provider.transcription_app_id) && !provider.chat_model) return "asr";
  return "llm";
}

export function hasConfiguredLlmProvider(providers: ApiProvider[]) {
  return providers.some((provider) =>
    providerCapability(provider) === "llm"
    && provider.is_enabled
    && Boolean(provider.base_url)
    && Boolean(provider.chat_model)
    && Boolean(provider.api_key_configured)
  );
}

export function hasConfiguredAsrProvider(providers: ApiProvider[]) {
  return providers.some((provider) =>
    providerCapability(provider) === "asr"
    && provider.is_enabled
    && Boolean(provider.transcription_app_id)
    && Boolean(provider.transcription_model)
    && Boolean(provider.transcription_access_token_configured)
    && Boolean(provider.transcription_secret_key_configured)
  );
}

File: apps/api/app/schemas/providers.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import ProviderType


class ProviderPresetEntry(BaseModel):
    provider_type: ProviderType
    provider_name: str


class ProviderCreateRequest(BaseModel):
    provider_name: str = Field(min_length=1)
    provider_type: ProviderType
    capability: str | None = None
    base_url: str | None = None
    api_key: str | None = None
    chat_model: str | None = None
    embedding_model: str | None = None
    transcription_model: str | None = None
    transcription_app_id: str | None = None
    transcription_access_token: str | None = None
    transcription_secret_key: str | None = None
    thinking_mode: str | None = None
    is_enabled: bool = True


class ProviderUpdateRequest(ProviderCreateRequest):
    pass


class ProviderDeleteResponse(BaseModel):
    id: str
    deleted: bool


class ProviderEntry(BaseModel):
    id: str
    provider_name: str
    provider_type: ProviderType
    capability: str = "llm"
    base_url: str | None = None
    api_key_configured: bool = False
    chat_model: str | None = None
    embedding_model: str | None = None
    transcription_model: str | None = None
    transcription_app_id: str | None = None
    transcription_access_token_configured: bool = False
    transcription_secret_key_configured: bool = False
    thinking_mode: str = "default"
    is_enabled: bool = True
    last_test_status: str | None = None
    last_tested_at: datetime | None = None


class ProviderListResponse(BaseModel):
    items: list[ProviderEntry]


class ProviderTestResponse(BaseModel):
    provider_id: str
    ok: bool
    latency_ms: int
    message: str | None = None
